core/org.openjdk.jmc.testlib/src/main/java/org/openjdk/jmc/flightrecorder/test/util/JfrGenerator.java

```diff
@@ -0,0 +1,158 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.openjdk.jmc.flightrecorder.test.util;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Objects;
+import java.util.TimeZone;
+
+/**
+ * Test utility to produce a JFR recording for a simple java program. This allows us to validate
+ * that JMC rules catch problems in a specific program in a way that's easy for developers to
+ * follow, without checking in recording artifacts. These should be used judiciously because
+ * executing these processes and capturing recordings while running tests is more expensive than
+ * consuming stored recording data, and only provides feedback on JFRs from the active JDK version.
+ */
+public final class JfrGenerator {
+
+	public static SourceStage create() {
+		return new SourceStage();
+	}
+
+	public static final class SourceStage {
+		private SourceStage() {
+		}
+
+		/**
+		 * Provide source code as a string for a valid <a href="https://openjdk.org/jeps/330">JEP
+		 * 330</a> single-file source-code program, which will be profiled using JFR.
+		 */
+		public ConfigurationStage source(String source) {
+			return new ConfigurationStage(Objects.requireNonNull(source, "Source is required"));
+		}
+	}
+
+	public static final class ConfigurationStage {
+		private final String source;
+
+		private ConfigurationStage(String source) {
+			this.source = source;
+		}
+
+		/**
+		 * Name of a default configuration, for example {@code default} and {@code profile} are
+		 * included in most OpenJDK builds.
+		 */
+		public ExecutionStage configurationName(String configuration) {
+			return new ExecutionStage(source, Objects.requireNonNull(configuration, "Configuration is required"), null);
+		}
+
+		/**
+		 * Contents of a JFR configuration file.
+		 */
+		public ExecutionStage configuration(String configuration) {
+			return new ExecutionStage(source, null, Objects.requireNonNull(configuration, "Configuration is required"));
+		}
+	}
+
+	public static final class ExecutionStage {
+
+		private final String source;
+		private final String configurationName;
+		private final String configurationContents;
+
+		private ExecutionStage(String source, String configurationName, String configurationContents) {
+			this.source = source;
+			this.configurationName = configurationName;
+			this.configurationContents = configurationContents;
+		}
+
+		/**
+		 * Executes the provided Java source with the specified recording configuration, then
+		 * invokes {@code consumer} with the resulting recording.
+		 */
+		public void execute(RecordingConsumer consumer) throws IOException {
```

Comment:
Executing another JVM, to dynamically get recordings, seems both a bit risky and error prone to do as part of the testing. We assume that all the events have already been tested as part of the JDK testing, so this testing will be to ensure that we get the right output from the rule, given a certain input recording. 

I would suggest one of two ways of doing this:
Either simply make or generate a little recording, or some recording, that show the behaviour you are looking for. You can generate them, for example, by adding to https://github.com/thegreystone/recgen. Using the writer to generate them will result in recordings with exactly the shape you want, and they will also be minimal. Though, perhaps a bit unrealistic perhaps.

If you want to generate them on the fly, you can do what recgen does, and do it from within the tests themselves.